src/scheduler/MpiWorld.cpp

```diff
@@ -170,13 +177,43 @@ void MpiWorld::create(const faabric::Message& call, int newId, int newSize)
 
 void MpiWorld::destroy()
 {
-    // Destroy once per host
-    if (!isDestroyed.test_and_set()) {
-        shutdownThreadPool();
+    // Destroy once per thread the rank-specific data structures
+    // Remote message endpoints
+    if (!mpiMessageEndpoints.empty()) {
+        for (auto& e : mpiMessageEndpoints) {
+            if (e != nullptr) {
+                e->close();
+            }
+        }
+        mpiMessageEndpoints.clear();
+    }
 
-        // Note - we are deliberately not deleting the KV in the global state
-        // TODO - find a way to do this only from the master client
+    // Unacked message buffers
+    if (!unackedMessageBuffers.empty()) {
+        for (auto& umb : unackedMessageBuffers) {
+            if (umb != nullptr) {
+                if (!umb->isEmpty()) {
+                    SPDLOG_ERROR("Destroying the MPI world with outstanding {}"
+                                 " messages in the message buffer",
+                                 umb->size());
+                    throw std::runtime_error(
+                      "Destroying world with a non-empty MPI message buffer");
+                }
+            }
+        }
+        unackedMessageBuffers.clear();
+    }
```

Comment:
Would be good to add a couple of small tests to make sure these `destroy` checks work, e.g. call `isend` a few times then call `destroy` and `REQUIRE_THROWS`

Reply:
Thanks for pointing out, as I had actually missed checking if `iSendRequest` was empty at destruction time.